page-index-types

joe-ucp opened the pull request:
Goal
Introduce `Vec<Vec<Option<...>>>` for column and offset page indexes and update the parquet crate so it compiles and existing behavior continues to work, while keeping semantics changes as small and explicit as possible.

Which issue does this PR close?
Part of #8818.

Rationale for this change
The Parquet format allows page indexes (column index and offset index) to be optional per column chunk. The existing Rust types:

- `pub type ParquetColumnIndex = Vec<Vec<ColumnIndexMetaData>>`
- `pub type ParquetOffsetIndex = Vec<Vec<OffsetIndexMetaData>>`

implicitly assume that once page indexes are loaded, every column chunk has an index. This makes it awkward (and sometimes impossible) to represent partially missing indexes and encourages `unwrap()`-based code paths.

This PR introduces an `Option` at the leaf of these containers so we can explicitly represent “no index for this (row_group, column)” and prepare the codebase for safer semantics in follow-up PRs.

What changes are included in this PR?
- Type alias changes

  - Change the page index aliases to:

    ```rust
    pub type ParquetColumnIndex = Vec<Vec<Option<ColumnIndexMetaData>>>;
    pub type ParquetOffsetIndex = Vec<Vec<Option<OffsetIndexMetaData>>>;
    ```

  - Clarify the documentation comments for these aliases to mention `None` at the leaf as “no index for this column chunk”.

- Minimal plumbing to keep existing code compiling

  - Update internal metadata builders and writers to construct `ParquetColumnIndex` and `ParquetOffsetIndex` with `Some(...)` where indexes are currently generated.
  - Update call sites that indexed into `[row_group][column]` and previously saw a bare `ColumnIndexMetaData` or `OffsetIndexMetaData` to accept an `Option<...>`. In this PR we mostly:
    - Wrap existing values in `Some(...)`, and
    - Where behavior previously assumed an index was always present, use `as_ref().expect(...)` to preserve the invariant and make the assumptions explicit in error messages.

- No behavioral logic changes yet

  - This PR intentionally avoids changing how missing indices are *interpreted* (for example, `ColumnIndexMetaData::NONE` is preserved where it was used before).
  - More nuanced handling of `None` vs `ColumnIndexMetaData::NONE`, and graceful behavior for partially missing indices, is left for later PRs.

## Are these changes tested?

Yes.

- Existing parquet tests pass with the new types and plumbing in place:
  ```bash
  cargo test --package parquet --lib
  ```
- No new behavior-focused tests are introduced in this PR; the intent is to introduce the new in-memory representation with minimal semantic changes and rely on existing tests to confirm that behavior is preserved.

---

## Are there any user-facing changes?

Yes, at the type level.

- **Public aliases:**
  - `ParquetColumnIndex` is now `Vec<Vec<Option<ColumnIndexMetaData>>>`
  - `ParquetOffsetIndex` is now `Vec<Vec<Option<OffsetIndexMetaData>>>`

- **Downstream code** that accessed these aliases as `indexes[row_group][column]` will need to handle `Option`, for example:

  ```rust
  if let Some(idx) = indexes[rg][col].as_ref() {
      // use idx
  }
  ```

- The Parquet on-disk format and footer encoding are **unchanged**.

Semantically, the goal of this PR is “same behavior, new types”; more meaningful changes in how missing indexes are represented and handled are done in follow-up PRs.

joe-ucp committed:
page-index-types

Goal: Introduce Vec<Vec<Option<...>>> for column & offset indexes and wire it up just enough so everything compiles and tests pass, without changing semantics more than necessary.

joe-ucp committed:
Make Parquet page indexes Vec<Vec<Option<...>>> and update parquet crate call sites so partial / missing page indexes are representable without changing behavior

etseidl commented on parquet/src/bin/parquet-fromcsv.rs:
```diff
@@ -445,6 +445,9 @@ mod tests {
         let pos = actual.find('\n').unwrap() + 1;
         actual = actual[pos..].to_string();
+        // Normalize line endings for cross-platform compatibility
+        let expected = expected.replace("\r\n", "\n").trim_start().to_string();
```
Also not relevant

joe-ucp replied:
Reverted. Restored the original logic to keep this PR scoped to page index typing only.

etseidl commented on parquet/src/file/metadata/reader.rs:
```diff
@@ -96,7 +96,7 @@ pub enum PageIndexPolicy {
     fn from(value: bool) -> Self {
         match value {
-            true => Self::Required,
+            true => Self::Optional,
```
This is probably preferable, but I think warrants some discussion. This should likely be in a different PR with its own issue.

joe-ucp replied:
Reverted to pre-PR behavior. I’ll open a separate issue/PR if we want to revisit the boolean-to-policy mapping. For this PR, no change from legacy behavior.

etseidl commented on parquet/src/file/metadata/writer.rs:
```diff
@@ -181,16 +170,8 @@ impl<'a, W: Write> ThriftMetadataWriter<'a, W> {
-            })
-        };
+        let offset_indexes: Option<ParquetOffsetIndex> =
+            if all_none { None } else { offset_indexes };
```
Same comment here

joe-ucp replied:
### Offset Index Path Return Value Update

The offset index path now also returns:

```rust
Ok(if all_none { None } else { offset_indexes })
```

- If `all_none` is `true`, returns `Ok(None)`.
- Otherwise, returns `Ok(offset_indexes)`.

This provides consistency and clarity in the return values.

etseidl commented on parquet/src/file/metadata/writer.rs:
```diff
@@ -487,10 +468,7 @@ impl<'a, W: Write> ParquetMetaDataWriter<'a, W> {
-                            .iter()
-                            .map(|column_index| Some(column_index.clone()))
-                            .collect()
+                        column_indexes.to_vec()
```
This whole function is no longer needed. The calls to `convert_column_index()` and `convert_offset_index()` above can instead be 
```rust
      let column_indexes = self.metadata.column_index();
      let offset_indexes = self.metadata.offset_index();
```

joe-ucp replied:
### Helper Removal and Direct Metadata Usage

- Removed the `convert_*` helper functions.
- Eliminated all `map(|x| Some(x.clone()))` loops.
- Now directly use:
  - `self.metadata.column_index()`
  - `self.metadata.offset_index()`
- Finalization functions are now pass-through with the `all_none` collapse logic preserved.

These changes simplify the code and improve clarity.

etseidl reviewed (changes requested):
Thanks @joe-ucp, I think this will be a change for the better. This is a first pass review.

There seem to be a number of trivial changes that are either not strictly necessary for this PR, or not desirable at all. Please remove these as they just distract from what is already a fairly large (by necessity) PR.

For the most part these changes seem correct, but there are transformations done in `file/metadata/writer.rs` that are no longer necessary now that we've changed the page indexes in the `ParquetMetaData` to match how they are produced.

Also be aware that as this is a breaking API change, we won't be able to merge it until `58.0.0` some time in January.

joe-ucp committed:
parquet: finalize page-index Option leaves; writer passthrough; policy restore

Introduce Vec<Vec<Option<...>>> for page indexes and keep semantics stable.

Revert unrelated edits in arrow/array_reader, arrow/decoder, arrow/schema/extension (unused-mut), and parquet-fromcsv.

Restore From<bool> for PageIndexPolicy to pre-PR behavior (true -> Required, false -> Skip).

Parser: do not enforce Required for column index; retain enforcement for offset index only.

Writer: remove conversion helpers and map(|x| Some(...)) loops; use self.metadata.column_index() / offset_index() directly; collapse to None when all chunks are None.

Update docs for ParquetColumnIndex/ParquetOffsetIndex with Option semantics and idiomatic access examples.

Keep on-disk/footer encoding unchanged; preserve ColumnIndexMetaData::NONE behavior.

Call sites updated to handle Option (.as_ref().expect(...) where legacy invariant held; if let Some(...) otherwise).

Breaking change (58.0.0):

Public aliases now include Option at the leaf:

ParquetColumnIndex = Vec<Vec<Option<ColumnIndexMetaData>>>

ParquetOffsetIndex = Vec<Vec<Option<OffsetIndexMetaData>>>

Migration:

if let Some(idx) = indexes[rg][col].as_ref() { /* use idx */ }

Refs: #8818 (part), follow-ups #8892, #8893

joe-ucp commented on parquet/src/file/metadata/parser.rs:
```diff
@@ -269,14 +269,26 @@ pub(crate) fn parse_column_index(
 
     metadata.set_column_index(Some(index));
+
+    if column_index_policy == PageIndexPolicy::Required {
```
Addressed. The PageIndexPolicy::Required check for column index has been removed. We retain enforcement for the offset index only, matching existing semantics and your guidance.

etseidl commented on parquet/src/arrow/array_reader/mod.rs:
```diff
@@ -51,7 +51,7 @@ mod test_util;
 // Note that this crate is public under the `experimental` feature flag.
 use crate::file::metadata::RowGroupMetaData;
-pub use builder::{ArrayReaderBuilder, CacheOptions, CacheOptionsBuilder};
+pub use builder::{ArrayReaderBuilder, CacheOptionsBuilder};
```
The original change removed `CacheOptions`, it didn't add it. Nothing has been reverted here.

etseidl reviewed (changes requested):
Some more superfluous changes to revert. There will also need to be some changes made to satisfy clippy. In the future please follow the instructions [here](https://github.com/apache/arrow-rs/blob/main/CONTRIBUTING.md#clippy-lints)

etseidl commented:
Switching to draft since this is not waiting on a review.
